### PR TITLE
Scaffolding: Annotate NRTs

### PR DESCRIPTION
**Pull request by bricelam**
Part of #19007

### CODE REVIEW

File: src/EFCore.Relational/Design/AnnotationCodeGenerator.cs [hunk 1]
```diff
@@ -655,7 +657,7 @@ protected virtual AttributeCodeFragment GenerateDataAnnotation([NotNull] IProper
             }
         }
 
-        private static bool TryGetAndRemove<T>(IDictionary<string, IAnnotation> annotations, string annotationName, out T annotationValue)
+        private static bool TryGetAndRemove<T>(IDictionary<string, IAnnotation> annotations, string annotationName, out T? annotationValue)
```
Comment (smitpatel) [inline src/EFCore.Relational/Design/AnnotationCodeGenerator.cs:660]: NotNullWhen on out T?

File: src/EFCore.SqlServer/Design/Internal/SqlServerAnnotationCodeGenerator.cs [hunk 2]
```diff
@@ -93,11 +95,11 @@ protected override bool IsHandledByConvention(IModel model, IAnnotation annotati
 
             if (annotation.Name == RelationalAnnotationNames.DefaultSchema)
             {
-                return (string)annotation.Value == "dbo";
+                return (string?)annotation.Value == "dbo";
             }
 
             return annotation.Name == SqlServerAnnotationNames.ValueGenerationStrategy
-                && (SqlServerValueGenerationStrategy)annotation.Value == SqlServerValueGenerationStrategy.IdentityColumn;
+                && (SqlServerValueGenerationStrategy)annotation.Value! == SqlServerValueGenerationStrategy.IdentityColumn;
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Design/Internal/SqlServerAnnotationCodeGenerator.cs:102]: Can the value be null? Casting to nullable may get same results.

File: src/EFCore.SqlServer/Design/Internal/SqlServerAnnotationCodeGenerator.cs [hunk 3]
```diff
@@ -175,8 +177,8 @@ protected override MethodCallCodeFragment GenerateFluentApi(IIndex index, IAnnot
                         (name, schema) switch
                         {
                             (null, null) => Array.Empty<object>(),
-                            (_, null) => new object[] { name },
-                            _ => new object[] { name, schema }
+                            (_, null) => new object[] { name! },
+                            _ => new object[] { name!, schema! }
```
Comment (roji) [inline src/EFCore.SqlServer/Design/Internal/SqlServerAnnotationCodeGenerator.cs:181]: Isn't the schema actually optional (defaulting to default model-level schema)? If so this should be `object?[]` no?

File: src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs [hunk 4]
```diff
@@ -111,7 +111,9 @@ public class SqlServerTypeMappingSource : RelationalTypeMappingSource
             = new SqlServerDoubleTypeMapping("float");
 
         private readonly SqlServerDateTimeOffsetTypeMapping _datetimeoffset
+#pragma warning disable CS8620 // TODO: Follow up. Possibly dotnet/roslyn#50311
```
Comment (roji) [inline src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs:114]: :+1:

**Review by roji — APPROVED** [review]

File: src/EFCore.Relational/Design/AnnotationCodeGenerator.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore.Relational/Design/AnnotationCodeGenerator.cs:660]: We return true only when annotation.Value is non-null.

File: src/EFCore.SqlServer/Design/Internal/SqlServerAnnotationCodeGenerator.cs (hunk 3 above)
Comment (smitpatel) [reply src/EFCore.SqlServer/Design/Internal/SqlServerAnnotationCodeGenerator.cs:181]: Case expressions, this part captures when name/schema both are non-null.

File: src/EFCore.SqlServer/Design/Internal/SqlServerAnnotationCodeGenerator.cs (hunk 2 above)
Comment (bricelam) [reply src/EFCore.SqlServer/Design/Internal/SqlServerAnnotationCodeGenerator.cs:102]: We never insert a `Nullable<SqlServerValueGenerationStrategy>` value for this key.

File: src/EFCore.Relational/Design/AnnotationCodeGenerator.cs (hunk 1 above)
Comment (bricelam) [reply src/EFCore.Relational/Design/AnnotationCodeGenerator.cs:660]: Fixed. (Call sites too)

**Review by smitpatel — APPROVED** [review]

**Pull request merged**